styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -472,20 +472,22 @@ void pollPods() {
   }
 
   private synchronized void tryPollPods() {
-    final Map<WorkflowInstance, RunState> activeStates = stateManager.getActiveStates();
-
+    final Set<WorkflowInstance> runningWorkflowInstances = getRunningWorkflowInstances();
     final PodList list = client.pods().list();
-    examineRunningWFISandAssociatedPods(activeStates, list);
+    examineRunningWFISandAssociatedPods(runningWorkflowInstances, list);
 
     for (Pod pod : list.getItems()) {
       logEvent(Watcher.Action.MODIFIED, pod, list.getMetadata().getResourceVersion(), true);
       final Optional<WorkflowInstance> workflowInstance = readPodWorkflowInstance(pod);
       if (!workflowInstance.isPresent()) {
         continue;
       }
-      final RunState runState = activeStates.get(workflowInstance.get());
-      if (runState != null && isPodRunState(pod, runState)) {
-        emitPodEvents(Watcher.Action.MODIFIED, pod, runState);
+      // do an extra lookup to get latest state
+      // this is crucial because after fetching all states in datastore, some states might
+      // have transitioned already
```

Comment:
We might even observe pods whose states where created after fetching the states.

Reply:
Yeah good catch. Could you help fix this? I’m afk now. :)